File: 0014_longest_common_prefix/binary_search/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cassert>

class Solution {
public:
    std::string longestCommonPrefix(std::vector<std::string>& strs) {
        if (strs.empty()) return "";
        
        int minLen = strs[0].length();
        for (const std::string& s : strs) {
            minLen = std::min(minLen, (int)s.length());
        }
        
        int low = 1, high = minLen;
        int ans = 0;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isCommonPrefix(strs, mid)) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return strs[0].substr(0, ans);
    }

private:
    bool isCommonPrefix(const std::vector<std::string>& strs, int len) {
        std::string prefix = strs[0].substr(0, len);
        for (int i = 1; i < strs.size(); i++) {
            if (strs[i].compare(0, len, prefix) != 0) {
                return false;
            }
        }
        return true;
    }
};

void test() {
    Solution sol;
    std::vector<std::string> strs1 = {"flower", "flow", "flight"};
    assert(sol.longestCommonPrefix(strs1) == "fl");
    std::cout << "All tests passed!" << std::endl;
}

int main() {
    test();
    return 0;
}

File: 0014_longest_common_prefix/binary_search/python/solution.py
from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        
        min_len = min(len(s) for s in strs)
        low = 1
        high = min_len
        ans = 0
        
        while low <= high:
            mid = (low + high) // 2
            if self.isCommonPrefix(strs, mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1
                
        return strs[0][:ans]

    def isCommonPrefix(self, strs: List[str], length: int) -> bool:
        prefix = strs[0][:length]
        for i in range(1, len(strs)):
            if not strs[i].startswith(prefix):
                return False
        return True
